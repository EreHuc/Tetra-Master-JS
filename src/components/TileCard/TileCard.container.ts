import { connect } from "react-redux";

import { getPlayer, getTile } from "../../store";
import { TileCard } from "./TileCard";

const mapStateToProps = (state, ownProps) => ({
  tile: getTile(ownProps.tileId)(state),
  player: getPlayer(ownProps.playerId)(state),
});

const enhance = connect(mapStateToProps);

// @ts-ignore: Still having some issues with Redux and Typescript
export default enhance(TileCard);
