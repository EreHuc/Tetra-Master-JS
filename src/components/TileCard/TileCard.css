.tile-card {
  width: 64px;
  height: 64px;
  border: 2px solid transparent;
}

.tile-card.selected {
  border-color: #000;
}

.tile-type-1 {
  background: lightblue;
}

.tile-type-2 {
  background: lightcoral
}

.tile-type-3 {
  background: lightyellow
}

.tile-type-4 {
  background: lightgreen
}
