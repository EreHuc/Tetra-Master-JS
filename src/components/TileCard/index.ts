export { default } from "./TileCard.container";
export * from "./TileCard";
