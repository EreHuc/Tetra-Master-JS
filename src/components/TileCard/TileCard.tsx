import * as React from "react";

import { Player, Tile } from "../../models";
import "./TileCard.css";

type TileProps = {
  tile: Tile;
  player: Player;
  focused?: boolean;
  selected?: boolean;
  onMouseEnter?: React.MouseEventHandler;
  onMouseLeave?: React.MouseEventHandler;
  onClick?: React.MouseEventHandler;
};

export const TileCard: React.SFC<TileProps> = ({
  tile,
  player,
  selected,
  focused,
  onMouseEnter,
  onMouseLeave,
  onClick,
}) => (
  <button
    className={`tile-card tile-type-${tile.typeId} ${selected &&
      "selected"} player-${player.id}`}
    onMouseEnter={onMouseEnter}
    onMouseLeave={onMouseLeave}
    onClick={onClick}
  >
    {focused && `tile ${tile.id}`}
  </button>
);
