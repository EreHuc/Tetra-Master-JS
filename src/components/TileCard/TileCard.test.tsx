import { shallow } from "enzyme";
import * as React from "react";

import { TileCard } from "./TileCard";

const mockTile = { id: "1", typeId: "1" };
const mockPlayer = { id: "1", name: "Player 1", hand: [] };

describe("TileCard", () => {
  it("should not display the tile infos", () => {
    const wrapper = shallow(<TileCard tile={mockTile} player={mockPlayer} />);

    expect(wrapper.contains("tile 1")).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should have a className containing the tileId", () => {
    const wrapper = shallow(<TileCard tile={mockTile} player={mockPlayer} />);

    expect(wrapper.hasClass("tile-type-1")).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should have a className containing the playerId", () => {
    const wrapper = shallow(<TileCard tile={mockTile} player={mockPlayer} />);

    expect(wrapper.hasClass("player-1")).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should react to mouse events", () => {
    const onClick = jest.fn();
    const onMouseEnter = jest.fn();
    const onMouseLeave = jest.fn();
    const wrapper = shallow(
      <TileCard
        tile={mockTile}
        player={mockPlayer}
        onClick={onClick}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
      />,
    );

    wrapper.find("button").simulate("click");
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith();

    wrapper.find("button").simulate("mouseEnter");
    expect(onMouseEnter).toHaveBeenCalledTimes(1);
    expect(onMouseEnter).toHaveBeenCalledWith();

    wrapper.find("button").simulate("mouseLeave");
    expect(onMouseLeave).toHaveBeenCalledTimes(1);
    expect(onMouseLeave).toHaveBeenCalledWith();

    expect(wrapper).toMatchSnapshot();
  });

  describe("when focused", () => {
    it("should display the tile infos", () => {
      const wrapper = shallow(
        <TileCard tile={mockTile} player={mockPlayer} focused />,
      );

      expect(wrapper.contains("tile 1")).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when selected", () => {
    it("should have a 'selected' className", () => {
      const wrapper = shallow(
        <TileCard tile={mockTile} player={mockPlayer} selected />,
      );

      expect(wrapper.hasClass("selected")).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
