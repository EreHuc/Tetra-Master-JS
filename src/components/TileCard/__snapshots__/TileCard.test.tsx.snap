// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileCard should have a className containing the playerId 1`] = `
<button
  className="tile-card tile-type-1 undefined player-1"
/>
`;

exports[`TileCard should have a className containing the tileId 1`] = `
<button
  className="tile-card tile-type-1 undefined player-1"
/>
`;

exports[`TileCard should not display the tile infos 1`] = `
<button
  className="tile-card tile-type-1 undefined player-1"
/>
`;

exports[`TileCard should react to mouse events 1`] = `
<button
  className="tile-card tile-type-1 undefined player-1"
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
    }
  }
  onMouseEnter={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
    }
  }
  onMouseLeave={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
    }
  }
/>
`;

exports[`TileCard when focused should display the tile infos 1`] = `
<button
  className="tile-card tile-type-1 undefined player-1"
>
  tile 1
</button>
`;

exports[`TileCard when selected should have a 'selected' className 1`] = `
<button
  className="tile-card tile-type-1 selected player-1"
/>
`;
